text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+have arbitrary side effects. However, it is precisely that flexibility
+which leads to the two downsides cited above:
+
+1. The borrow checker must be more conservative because it does not,
+   and cannot, know what that two accessors like `children()` and
+   `set_position()` affect disjoint sets of fields.<sup>[1](#endnote1)</sup>
+2. The code generated must use a virtual call because it does not,
+   and cannot, know how any particular object plans to implement
+   `set_children()`.
+
+In a truly generic trait intended for implementation by arbitrary
+other types, this tradeoff for maximum flexibility is probably still a
+good idea. However, there are many cases where the flexibility to make
+accessors do arbitrary things (as opposed to simply access fields) is
+really not needed. Examples include traits that are local to a project
+(as opposed to public interfaces) as well as
+[attempts to model class-like hierarchies][aturon]. It seems best to
+leave this decision in the hands of the trait designer.
+
+[aturon]: http://aturon.github.io/blog/2015/09/18/reuse/
+
+# Detailed design
+[design]: #detailed-design
+
+### Fields in traits
+
+We extend traits with an optional "field block", which has the same
+structure as the contents of a struct (that is, a comma-separated list
+of fields; this RFC also adds the notation of a embedded prefix,
+described below). If following by further items, the Field block is
+terminated by a semicolon, but otherwise the semicolon may be omitted
+if the field block is the last item in the trait.
+
+```rust
+trait Trait {
+    field1: Type1, // <-- fields within a block separated by commas.
+    field2: Type2  // <-- no semicolon needed if there are no more items
+}
+
+trait Trait {
+    field1: Type1,
+    field2: Type2; // <-- semicolon terminates the field block
+
+    fn foo();
+}
+```
+
+**NB.** It might be simpler to just require semicolons after fields, or else
+require that they appear at the end of a trait definition. The intention was
+that you should be able to copy-and-paste the contents of a struct into a trait.
+
+### Impls with fields
+
+Impls are also extended with a field block, which maps field names to
+lvalues:
+
+```rust
+impl Trait for Type {
+    field1: self.foo.bar,
+    field2: self.baz       // <-- optional terminating semicolon, as before
+}
+```
+
+Trait fields cannot be mapped to arbitrary lvalues. Rather, the lvalue
+expression must be an expression of the form `self(.F)*` where `F`
+represents some field name (possibly a fully qualified one, see
+below). In other words, `self.a.b.c` would be legal, but
+`(*self.a.b).c` would not; nor would `self.a[b]`. Furthermore, the
+lvalue expression must not contain any implicit derefs.
+
+These rules serve to ensure the following properties:
+
+1. the lvalue is located within the `self` value with some fixed pointer offset;
+2. borrowck can easily evaluate whether two field mappings are disjoint.
```

Comment:
Presumably, the types must match too? If the types have lifetimes, do we allow subtyping? Intuition tells me that since the fields are mutable, then they must be invariant thus no subtyping, but perhaps not since mut should imply unique.